timemachine/cpp/src/kernels/k_integrator.cuh

```diff
@@ -1,47 +1,51 @@
 #include "k_fixed_point.cuh"
 
-template <typename RealType>
-__global__ void update_forward_baoab(
+template <typename RealType, int D>
+__global__ void k_update_forward_baoab(
     const int N,
-    const int D,
     const RealType ca,
-    const unsigned int *__restrict__ idxs, // N
-    const RealType *__restrict__ cbs,      // N
-    const RealType *__restrict__ ccs,      // N
-    const RealType *__restrict__ noise,    // N x 3
-    RealType *__restrict__ x_t,
-    RealType *__restrict__ v_t,
-    const unsigned long long *__restrict__ du_dx,
+    const unsigned int *__restrict__ idxs,        // N
+    const RealType *__restrict__ cbs,             // N
+    const RealType *__restrict__ ccs,             // N
+    const RealType *__restrict__ noise,           // N x 3
+    RealType *__restrict__ x_t,                   // N x 3
+    RealType *__restrict__ v_t,                   // N x 3
+    const unsigned long long *__restrict__ du_dx, // N x 3
     const RealType dt) {
 
     int kernel_idx = blockIdx.x * blockDim.x + threadIdx.x;
     if (kernel_idx >= N) {
         return;
     }
-    int atom_idx;
-    if (idxs) {
-        atom_idx = idxs[kernel_idx];
-    } else {
-        atom_idx = kernel_idx;
-    }
+    int atom_idx = idxs == nullptr ? kernel_idx : idxs[kernel_idx];
+
     if (atom_idx >= N) {
         return;
     }
 
-    int d_idx = blockIdx.y;
-    int local_idx = atom_idx * D + d_idx;
-
-    RealType force = -FIXED_TO_FLOAT<RealType>(du_dx[local_idx]);
-
     // BAOAB (https://arxiv.org/abs/1203.5428), rotated by half a timestep
 
     // ca assumed to contain exp(-friction * dt)
     // cbs assumed to contain dt / mass
     // ccs assumed to contain sqrt(1 - exp(-2 * friction * dt)) * sqrt(kT / mass)
-    RealType v_mid = v_t[local_idx] + cbs[atom_idx] * force;
+    RealType atom_cbs = cbs[atom_idx];
+    RealType atom_ccs = ccs[atom_idx];
+
+    RealType force_x = -FIXED_TO_FLOAT<RealType>(du_dx[atom_idx * D + 0]);
+    RealType force_y = -FIXED_TO_FLOAT<RealType>(du_dx[atom_idx * D + 1]);
+    RealType force_z = -FIXED_TO_FLOAT<RealType>(du_dx[atom_idx * D + 2]);
+
+    RealType v_mid_x = v_t[atom_idx * D + 0] + atom_cbs * force_x;
+    RealType v_mid_y = v_t[atom_idx * D + 1] + atom_cbs * force_y;
+    RealType v_mid_z = v_t[atom_idx * D + 2] + atom_cbs * force_z;
+
+    v_t[atom_idx * D + 0] = ca * v_mid_x + atom_ccs * noise[atom_idx * D + 0];
+    v_t[atom_idx * D + 1] = ca * v_mid_y + atom_ccs * noise[atom_idx * D + 1];
+    v_t[atom_idx * D + 2] = ca * v_mid_z + atom_ccs * noise[atom_idx * D + 2];
 
-    v_t[local_idx] = ca * v_mid + ccs[atom_idx] * noise[local_idx];
-    x_t[local_idx] += 0.5 * dt * (v_mid + v_t[local_idx]);
+    x_t[atom_idx * D + 0] += static_cast<RealType>(0.5) * dt * (v_mid_x + v_t[atom_idx * D + 0]);
+    x_t[atom_idx * D + 1] += static_cast<RealType>(0.5) * dt * (v_mid_y + v_t[atom_idx * D + 1]);
+    x_t[atom_idx * D + 2] += static_cast<RealType>(0.5) * dt * (v_mid_z + v_t[atom_idx * D + 2]);
```

Comment:
Is this change to remove the parallelism over `D` what the PR description refers to as "remove strides"? If so, curious if you have an intuition for why this is faster than using `D` times as many threads. Is it something to do with scheduling overhead with too many threads?

Reply:
nit: I suppose this could still be written as a `for` loop and presumably would be unrolled by the optimizer?

Reply:
This has to do with memory locality. In this version we have each thread loading 3 floats that are directly next to each other allowing for more efficient memory caching than different blocks reading individual floats with gaps between the threads.

>  I suppose this could still be written as a for loop and presumably would be unrolled by the optimizer? (since D is now a template argument)

Perhaps? If we had it unrolled using `pragma unroll` we would have to share variables for the same value such `v_mid` which could (compiler might make separate registers) reduce the pipelining that can happen (can compute v_mid_x separate of v_mid_y, etc).  A good summary is here: https://dmacssite.github.io/materials/volkov10-GTC.pdf

Reply:
It is possible that the following
```
#pragma unroll 3
for (int i = 0 ; i < D; i++) {
    RealType v_mid = ....;
   x_t[atom_idx * D + i] = v_mid * v_t[atom_idx * D + i[;
}
```
Would generate a separate register per value of index, but that is not obvious and compiler dependent. Seems more explicit to have three registers (v_mid_x, v_mid_y, v_mid_z) that can be computed separately and used to compute the new x coordinates separately then to trust that the compiler is doing the right thing. 

Don't have a good understanding of how the unroll optimization works either, and if there wasn't an intermediate value that needs to be computed, unrolling would be the cleanest approach.

Reply:
I tested this by moving it into a loop with `#pragma unroll D` and it looks like the compiler ends up doing the 'wrong' thing. It generates a kernel that uses 28 registers rather than 38 registers, and we lose most of the performance gains. Even with the additional registers, the kernel stays at 100% occupancy on Cuda Arch 8.6 devices

Reply:
> The explicitness seems useful, but there seems to be an inconsistency between this style and D being a parameter, since the code will be incorrect for D other than 3.

Should we consider adding a static assertion that D is 3 wherever we manually unroll making this assumption?

Reply:
Added static assertion in https://github.com/proteneer/timemachine/pull/1109/commits/fd3b7a262797995f87b08399ef5555ae419ede17